feat: implement 0x52 - MSTORE Opcode

Quentash opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Resolves: #243 and #289

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

enitrat commented on crates/evm/src/instructions/memory_operations.cairo:
```diff
@@ -19,6 +20,10 @@ impl MemoryOperation of MemoryOperationTrait {
     /// Save word to memory.
     /// # Specification: https://www.evm.codes/#52?fork=shanghai
     fn exec_mstore(ref self: ExecutionContext) -> Result<(), EVMError> {
+        let offset: u32 = Into::<u256, Result<u32, EVMError>>::into((self.stack.pop()?))?;
```
can you use pop_n instead?

Quentash replied:
I'm now using pop_n as suggested but I noticed that this increase the gas cost of the operation.
Here is with 2 pop() :
```
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_only_F_offset_0 ... ok (gas usage est.: 550490)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_0xFF00_offset_1 ... ok (gas usage est.: 580980)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_only_F_offset_1 ... ok (gas usage est.: 869760)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_0xFF00_offset_0x20 ... ok (gas usage est.: 1539140)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_0xFF_offset_1 ... ok (gas usage est.: 1069230)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_1_offset_1 ... ok (gas usage est.: 869760)
```
Here is with pop_n() :
```
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_only_F_offset_0 ... ok (gas usage est.: 618740)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_only_F_offset_1 ... ok (gas usage est.: 938010)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_1_offset_1 ... ok (gas usage est.: 938010)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_0xFF_offset_1 ... ok (gas usage est.: 1137480)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_0xFF00_offset_1 ... ok (gas usage est.: 649230)
test evm::tests::test_instructions::test_memory_operations::test_exec_mstore_should_store_0xFF00_offset_0x20 ... ok (gas usage est.: 1607390)
```

This is not a huge difference (~8%) but I felt like mentionning it.

enitrat replied:
ah, interesting! thanks for letting me know, I will look into it

enitrat reviewed (commented):
It would be nice to add a tests that loads the entire memory and verifies the state of it. 

For example when you do

```rust
    ctx.stack.push(0xFF00);
    ctx.stack.push(0x01);
```

Then verifying that the memory is 64 bytes long, that the first word is 0xff and the second word is 0. 

You should also add two cases where we have: 
- no offset
- 32-bytes offset

Quentash commented:
Corrected "succeed" into "succeeded".
Used pop_n(2) instea of 2 pop().
Added a test case with offset 0.
Added a test case with offset 32.
Added memory length verification into tests.
Used integer::BoundedInt for max value.

ClementWalter commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_only_F_offset_0() {
```
`should_store_max_uint256_at_offset_0` as we don't really store "F"

ClementWalter commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_only_F_offset_1() {
```
`should_store_max_uint256_at_offset_1`

ClementWalter commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_1_offset_1() {
```
don't need this test

ClementWalter commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_0xFF_offset_1() {
```
don't need this test

ClementWalter commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_0xFF00_offset_1() {
```
don't need this test

ClementWalter commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_0xFF00_offset_0x20() {
```
don't need this test

ClementWalter reviewed (changes requested):
thx, there are too much test for this! in terms of unit testing, when writing tests for `exec_{}` we mainly need to test that we popped enough arguments, and that we pass them correctly to the underlying function implementing the logic (if the logic is in another funcion, as it is the case here).

Quentash commented on crates/evm/src/tests/test_instructions/test_memory_operations.cairo:
```diff
@@ -77,3 +78,123 @@ fn test_exec_pop_should_stack_underflow() {
+
+#[test]
+#[available_gas(20000000)]
+fn test_exec_mstore_should_store_1_offset_1() {
```
Not really, this test was imported from https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/VMTests/vmIOandFlowOperations/mstoreFiller.yml ( as many others that I will now remove since they indeed don't really have plus-value )

Quentash commented:
Removed many unnecessary tests.
Reverted back to single pop.
Replaced "only_Fs" by "max_uint256".
Added to self to-do list to review english past participle.
